dotrix_core/src/assets.rs

```diff
@@ -287,86 +337,56 @@ pub fn release(mut assets: Mut<Assets>, mut renderer: Mut<Renderer>) {
 /// Asset map getting trait
 pub trait AssetMapGetter<T> {
     /// Returns HashMap reference for selected asset type
-    fn map(&self) -> &HashMap<Id<T>, T>;
+    ///
+    /// Returns an option as the map may not have been made yet
+    /// and we don't have the mutable access required to make it
+    fn map(&self) -> Option<&HashMap<Id<T>, T>>;
     /// Returns mutable HashMap reference for selected asset type
     fn map_mut(&mut self) -> &mut HashMap<Id<T>, T>;
 
     /// Returns HashSet reference for selected asset type that have been removed
     /// since last clean up
-    fn map_removed(&self) -> Option<&HashSet<Id<T>>> {
-        None
-    }
+    fn map_removed(&self) -> Option<&HashSet<Id<T>>>;
     /// Returns mutable HashSet reference for selected asset type that have been removed
     /// since last clean up
-    fn map_removed_mut(&mut self) -> Option<&mut HashSet<Id<T>>> {
-        None
-    }
+    fn map_removed_mut(&mut self) -> &mut HashSet<Id<T>>;
 }
 
-impl AssetMapGetter<Animation> for Assets {
-    fn map(&self) -> &HashMap<Id<Animation>, Animation> {
-        &self.animations
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Animation>, Animation> {
-        &mut self.animations
-    }
-}
-
-impl AssetMapGetter<Texture> for Assets {
-    fn map(&self) -> &HashMap<Id<Texture>, Texture> {
-        &self.textures
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Texture>, Texture> {
-        &mut self.textures
-    }
-}
-
-impl AssetMapGetter<Mesh> for Assets {
-    fn map(&self) -> &HashMap<Id<Mesh>, Mesh> {
-        &self.meshes
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Mesh>, Mesh> {
-        &mut self.meshes
-    }
-}
-
-impl AssetMapGetter<Skin> for Assets {
-    fn map(&self) -> &HashMap<Id<Skin>, Skin> {
-        &self.skins
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Skin>, Skin> {
-        &mut self.skins
-    }
-}
-
-impl AssetMapGetter<Resource> for Assets {
-    fn map(&self) -> &HashMap<Id<Resource>, Resource> {
-        &self.resources
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Resource>, Resource> {
-        &mut self.resources
-    }
-}
-
-impl AssetMapGetter<Shader> for Assets {
-    fn map(&self) -> &HashMap<Id<Shader>, Shader> {
-        &self.shaders
-    }
-
-    fn map_mut(&mut self) -> &mut HashMap<Id<Shader>, Shader> {
-        &mut self.shaders
-    }
-
-    fn map_removed(&self) -> Option<&HashSet<Id<Shader>>> {
-        Some(&self.removed_shaders)
+impl<T: Asset> AssetMapGetter<T> for Assets {
+    fn map(&self) -> Option<&HashMap<Id<T>, T>> {
+        self.assets
+            .get(&TypeId::of::<T>())?
+            .as_any_ref()
+            .downcast_ref::<HashMap<Id<T>, T>>()
```

Comment:
I am thinking about some wrapper around the asset, that could solve generic issues, like marking an asset as dirty or deleted. What do you think?